File: ts/src/hub.ts
/**
 * ROAR Protocol — Hub server with federation support (TypeScript).
 * Mirrors python/src/roar_sdk/hub.py.
 *
 * Provides discovery registration, lookup, search and federation sync/export.
 *
 * Security properties:
 * - Challenge-response proof-of-possession registration (no TOFU)
 * - Challenge is one-time use (consume deletes) and expires quickly
 * - Body size cap (256 KiB)
 * - Fail-closed when identity cannot be verified
 */

import * as http from "http";
import * as crypto from "crypto";

import { AgentDirectory } from "./types.js";
import type { AgentCard, DiscoveryEntry } from "./types.js";
import { ChallengeStore } from "./hub_auth.js";

const MAX_BODY_BYTES = 256 * 1024;

function parseDiscoveryEntry(raw: unknown): DiscoveryEntry | null {
  if (typeof raw !== "object" || raw === null) return null;
  const e = raw as Record<string, unknown>;
  const card = e["agent_card"];
  if (typeof card !== "object" || card === null) return null;
  const identity = (card as Record<string, unknown>)["identity"];
  if (typeof identity !== "object" || identity === null) return null;
  if (typeof (identity as Record<string, unknown>)["did"] !== "string") return null;
  return {
    agent_card: card as AgentCard,
    registered_at: typeof e["registered_at"] === "number" ? e["registered_at"] : Date.now() / 1000,
    last_seen: typeof e["last_seen"] === "number" ? e["last_seen"] : Date.now() / 1000,
    hub_url: typeof e["hub_url"] === "string" ? e["hub_url"] : "",
  };
}

function json(res: http.ServerResponse, status: number, body: unknown): void {
  res.writeHead(status, { "Content-Type": "application/json" });
  res.end(JSON.stringify(body));
}

async function readBoundedJson(req: http.IncomingMessage): Promise<Record<string, unknown>> {
  return new Promise((resolve, reject) => {
    let body = "";
    let bytes = 0;

    req.on("data", (chunk: Buffer | string) => {
      const s = typeof chunk === "string" ? chunk : chunk.toString("utf8");
      bytes += Buffer.byteLength(s);
      if (bytes > MAX_BODY_BYTES) {
        req.destroy();
        reject(new Error("body_too_large"));
        return;
      }
      body += s;
    });

    req.on("end", () => {
      try {
        const parsed = JSON.parse(body || "{}") as Record<string, unknown>;
        resolve(parsed);
      } catch {
        reject(new Error("invalid_json"));
      }
    });
  });
}

function validateAgentCard(raw: unknown): AgentCard | null {
  if (typeof raw !== "object" || raw === null || Array.isArray(raw)) return null;
  const card = raw as any;
  if (typeof card.identity !== "object" || card.identity === null || Array.isArray(card.identity)) return null;
  if (typeof card.identity.did !== "string" || !card.identity.did) return null;
  // public_key may be null on some cards, but for hub registration we require one.
  if (typeof card.identity.public_key !== "string" || !card.identity.public_key) return null;
  if (!Array.isArray(card.identity.capabilities)) card.identity.capabilities = [];
  if (typeof card.description !== "string") card.description = "";
  if (!Array.isArray(card.skills)) card.skills = [];
  if (!Array.isArray(card.channels)) card.channels = [];
  if (typeof card.endpoints !== "object" || card.endpoints === null || Array.isArray(card.endpoints)) card.endpoints = {};
  if (!Array.isArray(card.declared_capabilities)) card.declared_capabilities = [];
  if (typeof card.metadata !== "object" || card.metadata === null || Array.isArray(card.metadata)) card.metadata = {};
  return card as AgentCard;
}

function verifyEd25519Nonce(publicKeyHex: string, signature: string, nonce: string): boolean {
  if (!signature.startsWith("ed25519:")) return false;
  try {
    const b64 = signature.slice("ed25519:".length);
    const rawSig = Buffer.from(b64, "base64url");

    const rawPub = Buffer.from(publicKeyHex, "hex");
    // SPKI DER wrapper for Ed25519 public key
    const spkiHeader = Buffer.from("302a300506032b6570032100", "hex");
    const der = Buffer.concat([spkiHeader, rawPub]);
    const pubKey = crypto.createPublicKey({ key: der, format: "der", type: "spki" });

    return crypto.verify(null, Buffer.from(nonce, "utf8"), pubKey, rawSig);
  } catch {
    return false;
  }
}

export interface ROARHubOptions {
  host?: string;
  port?: number;
  hub_id?: string;
  peer_urls?: string[];
}

export class ROARHub {
  private _host: string;
  private _port: number;
  private _hubUrl: string;
  private _directory: AgentDirectory;
  private _peers: string[];
  private _challenges: ChallengeStore;
  private _httpServer: http.Server | null = null;

  constructor(opts: ROARHubOptions = {}) {
    this._host = opts.host ?? "0.0.0.0";
    this._port = opts.port ?? 8090;
    this._hubUrl = opts.hub_id ?? `http://${this._host}:${this._port}`;
    this._directory = new AgentDirectory();
    this._peers = [...(opts.peer_urls ?? [])];
    this._challenges = new ChallengeStore();
  }

  get directory(): AgentDirectory {
    return this._directory;
  }

  addPeer(url: string): void {
    if (!this._peers.includes(url)) this._peers.push(url);
  }

  serve(): Promise<void> {
    return new Promise((resolve) => {
      this._httpServer = http.createServer(async (req, res) => {
        const method = req.method ?? "GET";
        const url = new URL(req.url ?? "/", `http://${req.headers.host ?? "localhost"}`);

        // GET /roar/health
        if (method === "GET" && url.pathname === "/roar/health") {
          json(res, 200, {
            status: "ok",
            protocol: "roar/1.0",
            hub_url: this._hubUrl,
            agents: this._directory.listAll().length,
            peers: this._peers.length,
          });
          return;
        }

        // POST /roar/agents/register
        if (method === "POST" && url.pathname === "/roar/agents/register") {
          try {
            const body = await readBoundedJson(req);
            const did = body["did"];
            const public_key = body["public_key"];
            const card = body["card"];

            if (typeof did !== "string" || !did) {
              json(res, 400, { error: "did is required" });
              return;
            }
            if (typeof public_key !== "string" || !public_key) {
              json(res, 400, { error: "public_key is required — cannot verify identity without key" });
              return;
            }
            if (typeof card !== "object" || card === null) {
              json(res, 400, { error: "card is required" });
              return;
            }

            let ch;
            try {
              ch = this._challenges.issue(did, public_key, card as Record<string, unknown>);
            } catch (e) {
              json(res, 503, { error: (e as Error).message });
              return;
            }

            json(res, 200, {
              challenge_id: ch.challenge_id,
              nonce: ch.nonce,
              expires_at: ch.expires_at,
            });
            return;
          } catch (e) {
            const msg = (e as Error).message;
            if (msg === "body_too_large") {
              json(res, 413, { error: "Request body too large" });
              return;
            }
            json(res, 400, { error: "Invalid JSON body" });
            return;
          }
        }

        // POST /roar/agents/challenge
        if (method === "POST" && url.pathname === "/roar/agents/challenge") {
          try {
            const body = await readBoundedJson(req);
            const challenge_id = body["challenge_id"];
            const signature = body["signature"];

            if (typeof challenge_id !== "string" || !challenge_id) {
              json(res, 400, { error: "challenge_id is required" });
              return;
            }
            if (typeof signature !== "string" || !signature) {
              json(res, 400, { error: "signature is required" });
              return;
            }

            const ch = this._challenges.consume(challenge_id);
            if (!ch) {
              json(res, 401, { error: "challenge_expired" });
              return;
            }

            const ok = verifyEd25519Nonce(ch.public_key, signature, ch.nonce);
            if (!ok) {
              json(res, 401, { error: "invalid_signature" });
              return;
            }

            // Register agent card (fail closed if invalid)
            const card = validateAgentCard(ch.card);
            if (!card) {
              json(res, 400, { error: "invalid card" });
              return;
            }

            const entry = this._directory.register(card);
            entry.hub_url = this._hubUrl;
            json(res, 200, { registered: true });
            return;
          } catch (e) {
            const msg = (e as Error).message;
            if (msg === "body_too_large") {
              json(res, 413, { error: "Request body too large" });
              return;
            }
            json(res, 400, { error: "Invalid JSON body" });
            return;
          }
        }

        // GET /roar/agents (optional capability)
        if (method === "GET" && url.pathname === "/roar/agents") {
          const capability = url.searchParams.get("capability");
          const entries = capability ? this._directory.search(capability) : this._directory.listAll();
          json(res, 200, { agents: entries.map((e) => e) });
          return;
        }

        // GET /roar/agents/search?capability=X
        if (method === "GET" && url.pathname === "/roar/agents/search") {
          const cap = url.searchParams.get("capability") ?? "";
          if (!cap) {
            json(res, 400, { error: "capability is required" });
            return;
          }
          json(res, 200, { agents: this._directory.search(cap) });
          return;
        }

        // GET /roar/agents/:did
        if (method === "GET" && url.pathname.startsWith("/roar/agents/")) {
          const did = decodeURIComponent(url.pathname.slice("/roar/agents/".length));
          if (!did) {
            json(res, 400, { error: "did is required" });
            return;
          }
          const entry = this._directory.lookup(did);
          if (!entry) {
            json(res, 404, { error: "Agent not found" });
            return;
          }
          json(res, 200, entry);
          return;
        }

        // POST /roar/federation/sync
        if (method === "POST" && url.pathname === "/roar/federation/sync") {
          try {
            const body = await readBoundedJson(req);
            const entriesRaw = body["entries"];
            if (!Array.isArray(entriesRaw)) {
              json(res, 400, { error: "entries must be a list" });
              return;
            }

            let imported = 0;
            for (const raw of entriesRaw) {
              try {
                const entry = parseDiscoveryEntry(raw);
                if (!entry) continue;
                const did = entry.agent_card.identity.did;
                // Don't overwrite locally-registered agents
                if (!this._directory.lookup(did)) {
                  // @ts-expect-error internal store access matches python behavior
                  this._directory._agents[did] = entry;
                  imported += 1;
                }
              } catch {
                // Skip invalid entries (fail closed per-entry)
              }
            }

            json(res, 200, { imported, total: entriesRaw.length });
            return;
          } catch (e) {
            const msg = (e as Error).message;
            if (msg === "body_too_large") {
              json(res, 413, { error: "Request body too large" });
              return;
            }
            json(res, 400, { error: "Invalid JSON body" });
            return;
          }
        }

        // GET /roar/federation/export
        if (method === "GET" && url.pathname === "/roar/federation/export") {
          const entries = this._directory.listAll();
          json(res, 200, {
            hub_url: this._hubUrl,
            exported_at: Date.now() / 1000,
            entries,
          });
          return;
        }

        json(res, 404, { error: "not_found" });
      });

      this._httpServer.listen(this._port, this._host, () => resolve());
    });
  }

  stop(): Promise<void> {
    return new Promise((resolve, reject) => {
      if (!this._httpServer) return resolve();
      this._httpServer.close((err) => (err ? reject(err) : resolve()));
    });
  }
}

File: python/src/roar_sdk/hub.py
# -*- coding: utf-8 -*-
"""ROAR Protocol — Hub server with federation support (Layer 2).

A ROAR Hub is a discovery server that:
  1. Maintains a registry of agent cards (AgentDirectory)
  2. Exposes a REST API for registration, lookup, and search
  3. Federates with other hubs by syncing entries (push/pull)

Hub API endpoints (mounted on a FastAPI app):

  POST /roar/agents/register   — begin challenge-response registration
  POST /roar/agents/challenge  — complete registration with signed proof
  DELETE /roar/agents/{did}    — unregister (requires signed proof)
  GET  /roar/agents            — list all agents
  GET  /roar/agents/{did}      — lookup single agent
  GET  /roar/agents/search?capability=X — search by capability
  POST /roar/federation/sync   — receive entries from a peer hub
  GET  /roar/federation/export — export all entries for a peer hub

Usage::

    from roar_sdk.hub import ROARHub

    hub = ROARHub(host="0.0.0.0", port=8090)
    hub.add_peer("https://hub2.example.com")
    hub.serve()
"""

from __future__ import annotations

import base64
import logging
import time
from typing import Any, List, Optional

from .hub_auth import ChallengeStore
from .types import AgentCard, AgentDirectory, DiscoveryEntry

# FastAPI / uvicorn are optional (server extra).  Import them at module level so
# that ``from __future__ import annotations`` does not turn type hints inside
# serve() into unresolvable forward-reference strings.
try:
    from fastapi import FastAPI, HTTPException, Request  # type: ignore[import]
    from fastapi.responses import JSONResponse  # type: ignore[import]
    import uvicorn  # type: ignore[import]
    _FASTAPI_AVAILABLE = True
except ImportError:
    _FASTAPI_AVAILABLE = False
    # Provide stubs so that type annotations inside serve() can still be written.
    # They are only evaluated at runtime when serve() is called, which will raise
    # ImportError before reaching the annotated functions anyway.
    Request = None  # type: ignore[assignment,misc]

logger = logging.getLogger(__name__)

# Maximum request body size: 256 KiB
_MAX_BODY_BYTES = 256 * 1024


class ROARHub:
    """A ROAR discovery hub with federation support.

    Requires: pip install 'roar-sdk[server]'
    """

    def __init__(
        self,
        host: str = "0.0.0.0",
        port: int = 8090,
        *,
        hub_id: str = "",
        peer_urls: Optional[List[str]] = None,
    ) -> None:
        self._host = host
        self._port = port
        self._hub_url = hub_id or f"http://{host}:{port}"
        self._directory = AgentDirectory()
        self._peers: List[str] = list(peer_urls or [])
        self._challenge_store = ChallengeStore()

    @property
    def directory(self) -> AgentDirectory:
        return self._directory

    def add_peer(self, hub_url: str) -> None:
        if hub_url not in self._peers:
            self._peers.append(hub_url)

    def serve(self) -> None:
        """Start the hub HTTP server.

        Requires: pip install 'roar-sdk[server]'
        """
        if not _FASTAPI_AVAILABLE:
            raise ImportError(
                "Hub server requires fastapi and uvicorn. "
                "Install them: pip install 'roar-sdk[server]'"
            )

        app = FastAPI(title=f"ROAR Hub — {self._hub_url}")
        hub = self

        # ── Helpers ───────────────────────────────────────────────────────

        async def _read_bounded_json(request: Request) -> dict:
            """Read request body with a 256 KiB cap, then parse as JSON."""
            body = await request.body()
            if len(body) > _MAX_BODY_BYTES:
                raise HTTPException(status_code=413, detail="Request body too large")
            import json as _json
            try:
                return _json.loads(body)
            except Exception:
                raise HTTPException(status_code=400, detail="Invalid JSON body")

        def _verify_ed25519(public_key_hex: str, signature_str: str, message: str) -> bool:
            """Verify an Ed25519 signature.

            *signature_str* must be of the form ``ed25519:<base64url>``.
            *message* is the UTF-8 plaintext that was signed.

            Raises ``ImportError`` if the cryptography package is absent.
            Raises ``ValueError`` for malformed inputs.
            Returns True on success, False on bad signature.
            """
            from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
            from cryptography.exceptions import InvalidSignature

            if not signature_str.startswith("ed25519:"):
                raise ValueError("Signature must start with 'ed25519:'")
            b64_part = signature_str[len("ed25519:"):]
            # base64url without padding
            padding = "=" * (-len(b64_part) % 4)
            try:
                sig_bytes = base64.urlsafe_b64decode(b64_part + padding)
            except Exception:
                raise ValueError("Malformed base64url in signature")

            try:
                pub_bytes = bytes.fromhex(public_key_hex)
            except ValueError:
                raise ValueError("Malformed public key hex")

            pub_key = Ed25519PublicKey.from_public_bytes(pub_bytes)
            try:
                pub_key.verify(sig_bytes, message.encode())
                return True
            except InvalidSignature:
                return False

        # ── Agent registration — step 1 ───────────────────────────────────

        @app.post("/roar/agents/register")
        async def register(request: Request):
            """Issue a challenge for proof-of-possession registration.

            Accepts: { "did": str, "public_key": str, "card": {...} }
            Returns: { "challenge_id": ..., "nonce": ..., "expires_at": ... }
            """
            body = await _read_bounded_json(request)

            did = body.get("did", "")
            public_key = body.get("public_key", "")
            card_raw = body.get("card", {})

            if not isinstance(did, str) or not did:
                return JSONResponse(
                    status_code=400,
                    content={"error": "did is required"},
                )
            if not isinstance(public_key, str) or not public_key:
                return JSONResponse(
                    status_code=400,
                    content={"error": "public_key is required — cannot verify identity without key"},
                )

            try:
                challenge = hub._challenge_store.issue(did, public_key, card_raw)
            except RuntimeError as exc:
                return JSONResponse(
                    status_code=503,
                    content={"error": str(exc)},
                )

            logger.info("Challenge issued for DID: %s  challenge_id: %s", did, challenge.challenge_id)
            return {
                "challenge_id": challenge.challenge_id,
                "nonce": challenge.nonce,
                "expires_at": challenge.expires_at,
            }

        # ── Agent registration — step 2 ───────────────────────────────────

        @app.post("/roar/agents/challenge")
        async def complete_challenge(request: Request):
            """Verify signed challenge and register the agent card.

            Accepts: { "challenge_id": str, "signature": "ed25519:<base64url>" }
            Returns: { "registered": true }
            """
            body = await _read_bounded_json(request)

            challenge_id = body.get("challenge_id", "")
            signature = body.get("signature", "")

            if not isinstance(challenge_id, str) or not challenge_id:
                return JSONResponse(status_code=400, content={"error": "challenge_id is required"})
            if not isinstance(signature, str) or not signature:
                return JSONResponse(status_code=400, content={"error": "signature is required"})

            # Consume challenge (deletes it — prevents replay)
            challenge = hub._challenge_store.consume(challenge_id)
            if challenge is None:
                return JSONResponse(status_code=401, content={"error": "challenge_expired"})

            # Verify signature
            try:
                valid = _verify_ed25519(challenge.public_key, signature, challenge.nonce)
            except ImportError:
                return JSONResponse(
                    status_code=503,
                    content={"error": "signature_verification_unavailable"},
                )
            except ValueError as exc:
                return JSONResponse(status_code=401, content={"error": f"invalid_signature: {exc}"})

            if not valid:
                return JSONResponse(status_code=401, content={"error": "invalid_signature"})

            # Register the card stored in the challenge
            try:
                card = AgentCard(**challenge.card)
            except Exception as exc:
                return JSONResponse(status_code=400, content={"error": f"invalid card: {exc}"})

            entry = hub._directory.register(card)
            entry.hub_url = hub._hub_url
            logger.info("Registered (challenge ok): %s", card.identity.did)
            return {"registered": True}

        # ── Agent unregistration — signed proof required ──────────────────

        @app.delete("/roar/agents/{did:path}")
        async def unregister(did: str, request: Request):
            """Unregister an agent. Requires a signed proof of ownership.

            Accepts: { "did": str, "signature": str, "nonce": str, "timestamp": float }
            The signature must cover the string: ``delete:{did}:{nonce}:{timestamp}``
            """
            body = await _read_bounded_json(request)

            signature = body.get("signature", "")
            nonce = body.get("nonce", "")
            timestamp = body.get("timestamp")

            if not signature or not nonce or timestamp is None:
                return JSONResponse(
                    status_code=400,
                    content={"error": "signature, nonce, and timestamp are required"},
                )

            try:
                timestamp = float(timestamp)
            except (TypeError, ValueError):
                return JSONResponse(status_code=400, content={"error": "timestamp must be a number"})

            # Reject stale requests
            if abs(time.time() - timestamp) > 60:
                return JSONResponse(status_code=401, content={"error": "timestamp_expired"})

            # Look up the registered entry to get the public key
            entry = hub._directory.lookup(did)
            if not entry:
                raise HTTPException(status_code=404, detail="Agent not found")

            public_key = entry.agent_card.identity.public_key
            if not public_key:
                return JSONResponse(
                    status_code=400,
                    content={"error": "registered card has no public_key — cannot verify delete request"},
                )

            message = f"delete:{did}:{nonce}:{timestamp}"
            try:
                valid = _verify_ed25519(public_key, signature, message)
            except ImportError:
                return JSONResponse(
                    status_code=503,
                    content={"error": "signature_verification_unavailable"},
                )
            except ValueError as exc:
                return JSONResponse(status_code=401, content={"error": f"invalid_signature: {exc}"})

            if not valid:
                return JSONResponse(status_code=401, content={"error": "invalid_signature"})

            hub._directory.unregister(did)
            logger.info("Unregistered (signed): %s", did)
            return {"status": "removed", "did": did}

        # ── Agent lookup ──────────────────────────────────────────────────

        @app.get("/roar/agents")
        async def list_agents(capability: Optional[str] = None):
            if capability:
                entries = hub._directory.search(capability)
            else:
                entries = hub._directory.list_all()
            return {"agents": [e.model_dump() for e in entries]}

        @app.get("/roar/agents/{did:path}")
        async def lookup_agent(did: str):
            entry = hub._directory.lookup(did)
            if not entry:
                raise HTTPException(status_code=404, detail="Agent not found")
            return entry.model_dump()

        # ── Federation ───────────────────────────────────────────────────

        @app.post("/roar/federation/sync")
        async def receive_sync(request: Request):
            """Accept a batch of DiscoveryEntry objects from a peer hub.

            Requires Authorization header with federation secret.
            """
            import hmac as _hmac
            fed_secret = _os.getenv("ROAR_FEDERATION_SECRET", "")
            if fed_secret:
                auth_header = request.headers.get("authorization", "")
                if not auth_header.startswith("Bearer ") or not _hmac.compare_digest(
                    auth_header[7:], fed_secret
                ):
                    raise HTTPException(status_code=401, detail="Invalid federation secret")
            else:
                logger.warning("ROAR_FEDERATION_SECRET not set — federation sync is unauthenticated")

            body = await _read_bounded_json(request)
            entries = body.get("entries", [])
            imported = 0
            for raw in entries:
                try:
                    entry = DiscoveryEntry(**raw)
                    # Don't overwrite locally-registered agents
                    if not hub._directory.lookup(entry.agent_card.identity.did):
                        hub._directory.register(entry.agent_card)
                        imported += 1
                except Exception as exc:
                    logger.warning("federation/sync: skipping entry: %s", exc)
            return {"imported": imported, "total": len(entries)}

        @app.get("/roar/federation/export")
        async def export_for_peers():
            """Export all local entries for peer hubs to import."""
            entries = hub._directory.list_all()
            return {
                "hub_url": hub._hub_url,
                "exported_at": time.time(),
                "entries": [e.model_dump() for e in entries],
            }

        @app.get("/roar/health")
        async def health():
            import os as _os
            redis_status = "not_configured"
            redis_url = _os.getenv("ROAR_REDIS_URL")
            if redis_url:
                try:
                    import redis as _redis
                    r = _redis.from_url(redis_url, socket_timeout=1.0, socket_connect_timeout=1.0)
                    r.ping()
                    redis_status = "connected"
                except Exception:
                    redis_status = "disconnected"
            return {
                "status": "healthy",
                "protocol": "roar/1.0",
                "hub_url": hub._hub_url,
                "agents": len(hub._directory.list_all()),
                "peers": len(hub._peers),
                "dependencies": {"redis": redis_status},
            }

        logger.info("ROAR Hub starting on http://%s:%d", self._host, self._port)
        uvicorn.run(app, host=self._host, port=self._port, log_level="warning")

    async def push_to_peers(self) -> dict:
        """Push local entries to all configured peer hubs.

        Requires: pip install 'roar-sdk[http]'
        """
        try:
            import httpx
        except ImportError:
            raise ImportError("Federation push requires httpx: pip install 'roar-sdk[http]'")

        entries = [e.model_dump() for e in self._directory.list_all()]
        payload = {
            "hub_url": self._hub_url,
            "exported_at": time.time(),
            "entries": entries,
        }
        results: dict[str, dict[str, Any]] = {}
        async with httpx.AsyncClient(timeout=10) as client:
            for peer in self._peers:
                try:
                    r = await client.post(f"{peer.rstrip('/')}/roar/federation/sync", json=payload)
                    results[peer] = {"status": r.status_code, "imported": r.json().get("imported", 0)}
                except Exception as exc:
                    results[peer] = {"status": "error", "detail": str(exc)}
        return results

    async def pull_from_peers(self) -> dict:
        """Pull entries from all configured peer hubs.

        Requires: pip install 'roar-sdk[http]'
        """
        try:
            import httpx
        except ImportError:
            raise ImportError("Federation pull requires httpx: pip install 'roar-sdk[http]'")

        results: dict[str, dict[str, Any]] = {}
        async with httpx.AsyncClient(timeout=10) as client:
            for peer in self._peers:
                try:
                    r = await client.get(f"{peer.rstrip('/')}/roar/federation/export")
                    data = r.json()
                    imported = 0
                    for raw in data.get("entries", []):
                        try:
                            entry = DiscoveryEntry(**raw)
                            did = entry.agent_card.identity.did
                            if not self._directory.lookup(did):
                                self._directory._agents[did] = entry
                                imported += 1
                        except Exception:
                            pass
                    results[peer] = {"imported": imported}
                except Exception as exc:
                    results[peer] = {"status": "error", "detail": str(exc)}
        return results
